src/System.Security.Cryptography.Cng/src/System/Security/Cryptography/CngKey.StandardProperties.cs

```diff
@@ -0,0 +1,302 @@
+        {
+            get
+            {
+                unsafe
+                {
+                    byte propertyValue;
+                    int cbResult;
+                    ErrorCode errorCode = Interop.NCrypt.NCryptGetProperty(_keyHandle, KeyPropertyName.ClrIsEphemeral, &propertyValue, sizeof(byte), out cbResult, CngPropertyOptions.CustomProperty);
+                    if (errorCode != ErrorCode.ERROR_SUCCESS)
+                    {
+                        // Third party Key providers, and Windows PCP KSP won't recognize this property;
+                        // and Win32 layer does not enforce error return contract.
+                        // Therefore, they can return whatever error code they think appropriate.
+                        return false;
+                    }
+
+                    if (cbResult != sizeof(byte))
+                        return false;
+
+                    if (propertyValue != 1)
+                        return false;
+
+                    return true;
+                }
+            }
+
+            private set
+            {
+                unsafe
+                {
+                    byte isEphemeral = value ? (byte)1 : (byte)0;
+                    ErrorCode errorCode = Interop.NCrypt.NCryptSetProperty(_keyHandle, KeyPropertyName.ClrIsEphemeral, &isEphemeral, sizeof(byte), CngPropertyOptions.CustomProperty);
+                    if (errorCode != ErrorCode.ERROR_SUCCESS)
+                        throw errorCode.ToCryptographicException();
+                }
+            }
+        }
+
+        /// <summary>
+        ///     Is this a machine key or a user key
+        /// </summary>
+        public bool IsMachineKey
+        {
+            get
+            {
+                CngKeyOpenOptions keyType = (CngKeyOpenOptions)_keyHandle.GetPropertyAsDword(KeyPropertyName.KeyType, CngPropertyOptions.None);
+                bool isMachineKey = (keyType & CngKeyOpenOptions.MachineKey) == CngKeyOpenOptions.MachineKey;
+                return isMachineKey;
+            }
+        }
+
+        /// <summary>
+        ///     The name of the key, null if it is ephemeral. We can only detect ephemeral keys created by
+        ///     the CLR. Other ephemeral keys, such as those imported by handle, will get a CryptographicException
+        ///     if they read this property.
+        /// </summary>
+        public string KeyName
+        {
+            get
+            {
+                if (IsEphemeral)
+                    return null;
+
+                string keyName = _keyHandle.GetPropertyAsString(KeyPropertyName.Name, CngPropertyOptions.None);
+                return keyName;
+            }
+        }
+
+        /// <summary>
+        ///     Size, in bits, of the key
+        /// </summary>
+        public int KeySize
+        {
+            get
+            {
+                int keySize = _keyHandle.GetPropertyAsDword(KeyPropertyName.Length, CngPropertyOptions.None);
+                return keySize;
```

Comment:
Why this pattern of store into a local and then returning the local, rather than just returning the result directly?  To help with debugging or something?